prover/src/composer/mod.rs

```diff
@@ -131,9 +141,40 @@ impl<E: FieldElement> DeepCompositionPoly<E> {
         // divide the composition polynomials by (x - z) and (x - z * g), respectively,
         // and add the resulting polynomials together; the output of this step
         // is a single trace polynomial T(x) and deg(T(x)) = trace_length - 2.
-        let trace_poly =
+        let mut trace_poly =
             merge_trace_compositions(vec![t1_composition, t2_composition], vec![self.z, next_z]);
 
+        // finally compose the final term associated to the Lagrange kernel trace polynomial if
+        // there is one present.
+        if let Some(poly) = trace_polys.lagrange_poly() {
+            let ood_eval_frame = ood_trace_states.lagrange_kernel_frame().expect(
+                "should contain OOD values for Lagrange kernel trace polynomial if we are here",
+            );
+
+            let log_trace_len = poly.len().ilog2();
+            let g = E::from(E::BaseField::get_root_of_unity(log_trace_len));
+            let mut xs = Vec::with_capacity(log_trace_len as usize + 1);
+
+            // push z
+            xs.push(self.z);
+
+            // compute the values (z * g), (z * g^2), (z * g^4), ..., (z * g^(2^(v-1)))
+            let mut g_exp = g;
+            for _ in 0..log_trace_len {
+                let x = g_exp * self.z;
+                xs.push(x);
+                g_exp *= g_exp;
+            }
+
+            let p_s = polynom::interpolate(&xs, ood_eval_frame.inner(), true);
+            let z_s = polynom::get_zero_roots(&xs);
+            let quotient = polynom::div(&polynom::sub(poly, &p_s), &z_s);
```

Comment:
This is the place where we'd potentially want to do a more optimized version of division, right?

Reply:
Yes, but I have actually went on and changed it to use a simple trick which is nothing but repeated synthetic division using Ruffini's rule. The idea is that we know the polynomial $z_S$ via all of its roots and thus can divide by it by dividing by its linear terms one by one.
I have ran some benchmarks and I am seeing the following improvements  

![poly_division_bench](https://github.com/facebook/winterfell/assets/82364884/56dbb7e4-89ff-4959-96f7-1c44c25264e1)

I have added the benchmark to the branch. Let me know if we should remove it instead.

Reply:
Nice! If I'm reading this correctly, the approach with synthetic division is about 4x faster than the naive approach - right? If so, that's a clear winner.

One potential thing to investigate in the future (maybe worth creating an issue for this): synthetic division is single threaded and so, for this part of the proof we would not be able to take advantage of many CPU cores. For small traces this probably doesn't matter too much, but for long traces, I wonder if FFT-based division would outperform this at some point as it can be parallelized.

Reply:
> Nice! If I'm reading this correctly, the approach with synthetic division is about 4x faster than the naive approach - right? If so, that's a clear winner.

That is indeed the case.

> One potential thing to investigate in the future (maybe worth creating an issue for this): synthetic division is single threaded and so, for this part of the proof we would not be able to take advantage of many CPU cores. For small traces this probably doesn't matter too much, but for long traces, I wonder if FFT-based division would outperform this at some point as it can be parallelized.

I have experimented with an FFT-based routine but it was half-way between the two methods above in terms of performance. The testing was done serially though so probably not representative.
One thing to keep in mind with an FFT-based routine is that it is not straightforward to make it work for any two polynomials (i.e., it is ill-conditioned). The reason being that the divisor can have zeros in its FFT representation. In our case, we know exactly the zeros of the divisor so it is not that big of a deal.
Another solution would be to separate all of the logic for building the Lagrange column contribution to the DEEP composition poly and investigate if the inherent serial-ness of the synthetic division method is still an issue then.

Reply:
Yep - definitely something to investigate in the future. I'd be surprised if the FFT-based approach would be more performant than synthetic division based approach when run on fewer than 8 cores in parallel - so, benchmarking would be much more involved here.